### PR TITLE
Add BotStateMachine (In Progress)

### PR DESCRIPTION
**Pull request by kendallc**

### CODE REVIEW

File: pybotframework/botstate.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            return self.current_state
+        else:
+            if bot_state_data.intent in next_states:
+                return bot_state_data.intent
+            else:
+                return None
+
+
+class BotStateData(object):
+    def __init__(self, user_message,
```
Comment (michhar) [inline pybotframework/botstate.py]: Perhaps, add a place for the conversation id to keep track of that association.

File: pybotframework/botstate.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+
+
+class BotStateMachine(object):
+
+    def __init__(self, states):
```
Comment (michhar) [inline pybotframework/botstate.py:5]: What are the possible states if they are discrete?  Are we going off of https://github.com/google/bottery/ Actions?

File: pybotframework/botstate.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+
+
+class BotStateMachine(object):
```
Comment (michhar) [inline pybotframework/botstate.py:3]: Nice way to keep compatibility with Python 2 (including "object").  Perhaps we should do this elsewhere you think?

File: pybotframework/botstate.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    def transition_state(self, bot_state_data):
+        """
+        Based on the intent and entities provided, choose the next state
+        """
+        next_states = self.states['current_state']['next_states']
+        if len(next_states) == 0:
+            return self.current_state
+        else:
+            if bot_state_data.intent in next_states:
+                return bot_state_data.intent
```
Comment (michhar) [inline pybotframework/botstate.py:17]: should you set `next_state` here as well?

```python
bot_state_data.next_state = bot_state_data.intent
return bot_state_data.next_state
```

File: pybotframework/botstate.py (hunk 3 above)
Comment (kendallc) [reply pybotframework/botstate.py:3]: Yeah, we should do this everywhere we have a class.

**Pull request closed without merging**